Issue #1192 Separate local persistence from file based persistence

rkavanap opened the pull request:
Refactor to separate local persistence service from file based persistence.

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,71 @@
+   *
+   * @param safeSpaceId prepared logical space on which the physical space
+   */
+  void createSafeSpace(String safeSpaceId) throws CachePersistenceException;
```
Please, let's not use String there but instead some opaque id with a specific type.

ljacomet commented on impl/src/main/java/org/ehcache/impl/internal/store/disk/OffHeapDiskStore.java:
```diff
@@ -327,6 +328,9 @@ public int rankAuthority(ResourceType<?> authorityResource, Collection<ServiceCo
       if (serviceProvider == null) {
         throw new NullPointerException("ServiceProvider is null in OffHeapDiskStore.Provider.");
       }
+      if (diskPersistenceService == null) {
```
That error here makes no sense - do the check in `start` instead

ljacomet commented on impl/src/main/java/org/ehcache/impl/persistence/DefaultFileBasedPersistenceService.java:
```diff
@@ -0,0 +1,262 @@
+ * Default implementation of the {@link FileBasedPersistenceService} which can be used explicitly when
+ * {@link org.ehcache.PersistentUserManagedCache persistent user managed caches} are desired.
+ *
+ * @author RKAV refactored for issue-1192
```
Drop `@author` tag - we are not using them

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,71 @@
+   *
+   * @return Identifier that can be used to identify the safe space.
+   */
+  String prepareSafeSpace(String owner, String name);
```
What exactly is the `owner` concept providing?

rkavanap replied:
In case there are both frs and disk (and maybe hybrid) under the same cache manager. So when I destroy one it does not destroy the other..

ljacomet commented on impl/src/main/java/org/ehcache/impl/persistence/DefaultFileBasedPersistenceService.java:
```diff
@@ -0,0 +1,262 @@
+    }
+  }
+
+  //////PRIVATE/////////
```
Please match code style - we do not use such separators

ljacomet commented:
Given the release of 3.1.0, this PR cannot be merged before master has been moved to 3.2.0-SNAPSHOT

albinsuresh commented on impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java:
```diff
@@ -16,74 +16,44 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultLocalPersistenceService.class);
 
-  private final ConcurrentMap<String, PersistenceSpace> knownPersistenceSpaces = new ConcurrentHashMap<String, PersistenceSpace>();
+  private final ConcurrentMap<String, SafeSpace> createdSpaces = new ConcurrentHashMap<String, SafeSpace>();
```
Is it really required to maintain this map internally? The user of this service should be maintaining the reference to the `SafeSpaceIdentifier` instances that it prepares? Isn't the `SafeSpaceIdentifier` just a logical representation of the `SafeSpace`? And the `SafeSpaceIdentifier` generation is idempotent using the same owner and identifier values, right? So I'm not sure if we need to maintain it here. Let the caller keep the references to the `SafeSpaceIdentifier` that it has prepared or prepare a new one which will provide the same logical `SafeSpaceIdentifier`. I'm viewing this `SafeSpaceIdentifier` as analogous to `File` which is just a logical representation of a physical file just like `SafeSpaceIdentifier` is for a `SafeSpace`. May be there is something that I'm missing.

albinsuresh commented on impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java:
```diff
@@ -153,363 +101,196 @@ public synchronized void stop() {
-  public boolean handlesResourceType(ResourceType<?> resourceType) {
-    return ResourceType.Core.DISK.equals(resourceType);
-  }
+  public SafeSpaceIdentifier prepareSafeSpace(String owner, String identifier) {
```
The name as _prepareSafeSpaceIdentifier_ or __createSafeSpaceIdentifier_, may be? Since it is just about creating the identifier? Just a personal opinion.

ljacomet commented on impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java:
```diff
@@ -153,363 +97,168 @@ public synchronized void stop() {
-      }
+  @Override
+  public synchronized void startForMaintenance(ServiceProvider<MaintainableService> serviceProvider) {
+    if (rootDirectory == null) {
```
Why is that not an error?

rkavanap replied:
yes. this should be an error. Infact, this check is not even necessary as this object will not exist if rootDirectory is null...

ljacomet commented on impl/src/main/java/org/ehcache/impl/persistence/DefaultLocalPersistenceService.java:
```diff
@@ -153,363 +97,168 @@ public synchronized void stop() {
-      throw new AssertionError("All JDKs must have SHA-1");
-    }
+  private SafeSpace createSafeSpaceLogical(String owner, String identifier) {
+    File ownerDirectory = new File(rootDirectory, owner);
```
Is owner considered _safe_ because it is provided by Ehcache services and not user input?

rkavanap replied:
Yes. owner is typically hardcoded, pre-defined values used by Ehcache persistable resource services only. It is to identify a root for a given service that is separate from other services,

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,64 @@
    */
-  FileBasedPersistenceContext createPersistenceContextWithin(PersistenceSpaceIdentifier<?> identifier, String name) throws CachePersistenceException;
-}
+  FileBasedPersistenceContext createContextWithin(SafeSpaceIdentifier safeSpaceId, String name) throws CachePersistenceException;
```
As discussed with @rkavanap this method does not belong here

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,64 @@
+   * @param name name of safe spaces
+   * @param verbose log more
+   */
+  void destroySafeSpace(String owner, String name, boolean verbose);
```
Given that the `SafeSpaceIdentifier` can be obtained without any file based action, this makes little sense as one can always invoke the overloaded method.

ljacomet commented:
Still having issues with the naming here.

`LocalPersistenceService` and `FileBasedPersistenceService` are names too close to each other.

The first one represents the ability to obtain a `FileBasedPersistenceContext` to store information on disk. The second one is tailored for the needs of `DISK` resource and a close companion to the `DiskStore.Provider`.

Should they be named: `FileBasedPersistenceService` and DiskResourceService`? any other proposals?

rkavanap commented:
Addressed @ljacomet comments into 2 separate commits. 1 commit has the renames and 1 commit has other changes Will squash into a single commit post final review. 
 
Btw, I have not renamed LocalPersistenceService, since FileBasedPersistenceService is now renamed to DiskResourceService that reduced the confusion that @ljacomet stated above. But if @ljacomet  still thinks LocalPeristenceService should be renamed, it can be done in another commit (which can also then rename FileBasedStateRepository to DiskResourceStateRepository)..

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,50 @@
-   * @param name name of the context to create
-   * @return a {@link FileBasedPersistenceContext}
+   * @param safeSpaceId prepared logical space on which the physical space
    */
```
missing javadoc on `name` param

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,50 @@
 
   /**
-   * Creates a new persistence context within the given space.
+   * Create the safe spaces in actual physical space for the owner.
```
Typo: Creates

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,50 @@
+   */
+  void destroyAll(String owner);
+
+  // identifier that also returns the root of the space
```
Let's have proper javadoc here on the interface and its method. Especially as there is a catch on the `File` return, you cannot assume it exists ... and you should not be the one creating it ...

ljacomet commented on impl/src/main/java/org/ehcache/config/builders/UserManagedCacheBuilder.java:
```diff
@@ -207,7 +207,7 @@ T build(ServiceLocator serviceLocator) throws IllegalStateException {
         throw new IllegalStateException("Persistent user managed caches must have an id set");
       }
-      LocalPersistenceService persistenceService = serviceLocator.getService(LocalPersistenceService.class);
+      DiskResourceService persistenceService = serviceLocator.getService(DiskResourceService.class);
```
lets rename the variable as well.

ljacomet commented on impl/src/main/java/org/ehcache/config/builders/UserManagedCacheBuilder.java:
```diff
@@ -296,8 +296,8 @@ public void close() throws Exception {
     if (persistent) {
-      LocalPersistenceService persistenceService = serviceLocator
-          .getService(LocalPersistenceService.class);
+      DiskResourceService persistenceService = serviceLocator
```
lets rename the variable as well

ljacomet commented on impl/src/main/java/org/ehcache/impl/internal/persistence/DefaultFileBasedPersistenceServiceFactory.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ *
+ */
+public class DefaultFileBasedPersistenceServiceFactory implements ServiceFactory<DiskResourceService> {
```
That needs a rename to match the renamed service

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/DiskResourceService.java:
```diff
@@ -0,0 +1,38 @@
+import org.ehcache.spi.persistence.PersistableResourceService;
+
+/**
+ * Service to provide file based persistence context to caches tied to disk based stores.
```
I would rephrase this class level doc (in place of the two paragraphs you have):

> Resource service handling file level operations for disk tiers.

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,63 @@
- * <P>
- * Will be used by caches with a disk store, whether or not the data should survive a program restart.
- * <P/>
+ * Service that provides unique local persistence spaces (and safe sub spaces) to any service that requires it
```
> unique local persistence spaces (and safe sub spaces)

changed to
> isolated persistence spaces

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,63 @@
+  SafeSpaceIdentifier createSafeSpaceIdentifier(String owner, String name);
+
+  /**
+   * Creates or finds the safe space represented by {@code safeSpaceId} in the underlying physical space.
```
> Creates or finds ...

changed to
> Creates if it does not exist ...

ljacomet commented on core/src/main/java/org/ehcache/core/spi/service/LocalPersistenceService.java:
```diff
@@ -17,22 +17,63 @@
    */
-  FileBasedPersistenceContext createPersistenceContextWithin(PersistenceSpaceIdentifier<?> identifier, String name) throws CachePersistenceException;
-}
+  interface SafeSpaceIdentifier {
```
Should we rename that do `FileSpaceIdentifier`? Not sure the `Safe` brings anything.

rkavanap replied:
Isn't SafeSpaceIdentifier consistent with the rest of the calls (such as createSafeSpaceId..etc). So should I rename every method in the interface to FileSpace from SafeSpace?.

Isn't FileSpace pleonastic? or another way to look at it is space for file (which is directory). Since the service does a bit more than providing a directory (such as locking to prevent accidental use of the same directory under different cache managers and validating and unique naming etc.) thought SafeSpace conveys the meaning better.

ljacomet commented on impl/src/main/java/org/ehcache/impl/internal/persistence/DefaultDiskResourceServiceFactory.java:
```diff
@@ -0,0 +1,38 @@
+import org.ehcache.impl.persistence.DefaultDiskResourceService;
+import org.ehcache.spi.service.ServiceCreationConfiguration;
+
+/**
```
Get rid of the empty javadoc block

rkavanap commented:
test this please

ljacomet commented on impl/src/main/java/org/ehcache/impl/persistence/DefaultDiskResourceService.java:
```diff
@@ -0,0 +1,274 @@
+   */
+  @Override
+  public void destroy(String name) throws CachePersistenceException {
+    if (persistenceService == null) {
```
So because it is null, you have no option to start it even if you wanted to. And that explains the ignored tests.
Let me think about this ...

ljacomet replied:
Let's go ahead with this merge but not close #1192 under which I want to address re-enabling the tests.
So go ahead, rebase and squash commits.

ljacomet replied:
@henri-tremblay May want to give it some thought. Because this is now decomposed in two services, the solution we came up with for 3.1.2 no longer works. However, we should stick with the public behaviour IMO. Just find a different implementation to fix the problem.
